File: packages/memory/src/consumer.ts
/**
 * @fileoverview Memory consumer implementation
 * @module @anyq/memory/consumer
 */

import {
  BaseConsumer,
  createMessage,
  type MessageHandler,
  type BatchMessageHandler,
  type SubscribeOptions,
  type HealthStatus,
  type IMessage,
  ConnectionError,
} from '@anyq/core';
import type { MemoryQueueConfig } from './config.js';
import { MemoryQueue, type StoredMessage } from './queue.js';
import { getQueue, registerQueue } from './registry.js';

/**
 * Default subscribe options
 */
const DEFAULT_SUBSCRIBE_OPTIONS: Required<SubscribeOptions> = {
  fromBeginning: false,
  fromTimestamp: undefined as unknown as Date,
  concurrency: 1,
  autoAck: true,
  batchSize: 10,
  batchTimeout: 1000,
};

/**
 * In-memory consumer implementation
 *
 * Consumes messages from an in-memory queue for testing.
 *
 * @example
 * ```typescript
 * const consumer = new MemoryConsumer({
 *   driver: 'memory',
 *   queueName: 'test-queue',
 * });
 *
 * await consumer.connect();
 * await consumer.subscribe(async (message) => {
 *   console.log('Received:', message.body);
 *   await message.ack();
 * });
 * ```
 */
export class MemoryConsumer<T = unknown> extends BaseConsumer<T> {
  private queue: MemoryQueue<T> | null = null;
  private dlq: MemoryQueue<unknown> | null = null;
  private readonly queueName: string;
  private readonly maxDeliveryAttempts: number;
  private pollInterval: ReturnType<typeof setInterval> | null = null;
  private running = false;
  private parkTimers: Set<ReturnType<typeof setTimeout>> = new Set();

  // Memory adapter supports scheduled delivery natively via setTimeout
  // re-enqueue (see parkMessage below).
  protected get supportsNativeDelay(): boolean {
    return true;
  }

  constructor(config: MemoryQueueConfig) {
    super(config);
    this.queueName = config.queueName ?? 'default';
    this.maxDeliveryAttempts = config.deadLetterQueue?.maxDeliveryAttempts ?? 3;

    // Setup DLQ if configured
    if (config.deadLetterQueue?.enabled) {
      const dlqName = config.deadLetterQueue.destination ?? `${this.queueName}-dlq`;
      let dlq = getQueue<unknown>(dlqName);
      if (!dlq) {
        dlq = new MemoryQueue<unknown>(dlqName);
        registerQueue(dlqName, dlq);
      }
      this.dlq = dlq;
    }
  }

  /**
   * Connect to the in-memory queue
   */
  async connect(): Promise<void> {
    if (this._connected) {
      return;
    }

    try {
      // Get or create the queue
      let queue = getQueue<T>(this.queueName);
      if (!queue) {
        const config = this.config as MemoryQueueConfig;
        queue = new MemoryQueue<T>(this.queueName, {
          maxMessages: config.maxMessages,
          maxAgeMs: config.maxAgeMs,
        });
        registerQueue(this.queueName, queue);
      }

      this.queue = queue;
      this._connected = true;
      this.logger.info('Memory consumer connected', { queue: this.queueName });
    } catch (error) {
      throw new ConnectionError(
        'Failed to connect to memory queue',
        error instanceof Error ? error : undefined
      );
    }

    // No-op for memory (native delay), but kept uniform across adapters.
    this.verifyParkPolicy();
  }

  /**
   * Disconnect from the queue
   */
  async disconnect(): Promise<void> {
    this.beginShutdown();
    if (!this._connected) {
      return;
    }

    this.running = false;
    if (this.pollInterval) {
      clearInterval(this.pollInterval);
      this.pollInterval = null;
    }

    for (const timer of this.parkTimers) {
      clearTimeout(timer);
    }
    this.parkTimers.clear();

    this.queue = null;
    this._connected = false;
    this.logger.info('Memory consumer disconnected');
  }

  /**
   * Native dead-letter for memory: route via the existing MemoryQueue.deadLetter()
   * path so DLQ headers (x-original-queue, x-death-reason, ...) are preserved.
   */
  protected override async deadLetterMessage(
    message: IMessage<T>,
    reason: string,
  ): Promise<void> {
    const stored = message.raw as StoredMessage<T> | undefined;
    if (this.queue && this.dlq && stored) {
      this.queue.deadLetter(stored.id, this.dlq, new Error(reason));
      this.logger.warn('Message dead-lettered', {
        messageId: message.id,
        reason,
      });
      return;
    }
    // No DLQ configured: drop without requeue.
    this.logger.warn('No DLQ configured; dropping message', {
      messageId: message.id,
      reason,
    });
    await message.nack(false);
  }

  /**
   * Native park for memory: re-enqueue after `delayMs` via setTimeout.
   * The message body is re-published to the same queue so it gets a fresh
   * delivery attempt counter for the next pickup.
   */
  protected override async parkMessage(
    message: IMessage<T>,
    delayMs: number,
  ): Promise<void> {
    const queue = this.queue;
    if (!queue) {
      this.logger.warn('parkMessage called without a connected queue', {
        messageId: message.id,
      });
      await message.nack(false);
      return;
    }

    // Remove the message from the processing map so it doesn't sit there.
    const stored = message.raw as StoredMessage<T> | undefined;
    if (stored) {
      queue.ack(stored.id);
    }

    const body = message.body;
    const key = message.key;
    const headers = { ...message.headers };

    const timer = setTimeout(() => {
      this.parkTimers.delete(timer);
      try {
        queue.enqueue(body, { key, headers });
      } catch (err) {
        this.logger.error('Failed to re-enqueue parked message', {
          messageId: message.id,
          error: err instanceof Error ? err.message : String(err),
        });
      }
    }, delayMs);

    this.parkTimers.add(timer);
    this.logger.debug('Message parked for redelivery', {
      messageId: message.id,
      delayMs,
    });
  }

  /**
   * Subscribe to messages
   */
  async subscribe(
    handler: MessageHandler<T>,
    options?: SubscribeOptions
  ): Promise<void> {
    if (!this.queue) {
      throw new ConnectionError('Consumer not connected');
    }

    const opts = { ...DEFAULT_SUBSCRIBE_OPTIONS, ...options };
    this.running = true;

    // Poll for messages
    const poll = async () => {
      if (!this.running || this._paused || !this.queue) {
        return;
      }

      const stored = this.queue.dequeue();
      if (!stored) {
        return;
      }

      const message = this.createWrappedMessage(stored);

      try {
        this.emit('message', message);
        await handler(message);

        if (opts.autoAck) {
          await message.ack();
        }
      } catch (error) {
        const err = error instanceof Error ? error : new Error(String(error));
        const result = await this.applyStrategy(message, err, () =>
          handler(message),
        );
        if (!result.handled) {
          this.logger.error('Error processing message', {
            messageId: stored.id,
            error: err.message,
          });

          this.emit('error', err);

          // Legacy behaviour: DLQ after maxDeliveryAttempts; otherwise requeue.
          if (stored.deliveryAttempt >= this.maxDeliveryAttempts && this.dlq) {
            this.queue.deadLetter(stored.id, this.dlq, err);
            this.logger.warn('Message sent to DLQ', {
              messageId: stored.id,
              attempts: stored.deliveryAttempt,
            });
          } else {
            await message.nack(true);
          }
        }
      }
    };

    // Start polling
    this.pollInterval = setInterval(poll, 10);
    this.logger.info('Subscribed to queue', { queue: this.queueName });
  }

  /**
   * Subscribe to message batches
   */
  async subscribeBatch(
    handler: BatchMessageHandler<T>,
    options?: SubscribeOptions
  ): Promise<void> {
    if (!this.queue) {
      throw new ConnectionError('Consumer not connected');
    }

    const opts = { ...DEFAULT_SUBSCRIBE_OPTIONS, ...options };
    this.running = true;

    const poll = async () => {
      if (!this.running || this._paused || !this.queue) {
        return;
      }

      const storedMessages = this.queue.dequeueBatch(opts.batchSize);
      if (storedMessages.length === 0) {
        return;
      }

      const messages = storedMessages.map((stored) =>
        this.createWrappedMessage(stored)
      );

      try {
        await handler(messages);

        if (opts.autoAck) {
          for (const message of messages) {
            await message.ack();
          }
        }
      } catch (error) {
        const err = error instanceof Error ? error : new Error(String(error));
        // Memory supports per-message ack, so apply the strategy per message.
        let anyUnhandled = false;
        for (const message of messages) {
          const result = await this.applyStrategy(message, err, () =>
            handler([message]),
          );
          if (!result.handled) {
            anyUnhandled = true;
          }
        }
        if (anyUnhandled) {
          this.logger.error('Error processing batch', {
            count: messages.length,
            error: err.message,
          });
          this.emit('error', err);
          for (const message of messages) {
            await message.nack(true);
          }
        }
      }
    };

    this.pollInterval = setInterval(poll, opts.batchTimeout);
    this.logger.info('Subscribed to queue (batch)', { queue: this.queueName });
  }

  /**
   * Pause consumption
   */
  async pause(): Promise<void> {
    await super.pause();
    this.running = false;
  }

  /**
   * Resume consumption
   */
  async resume(): Promise<void> {
    await super.resume();
    this.running = true;
  }

  /**
   * Get health status
   */
  async healthCheck(): Promise<HealthStatus> {
    return {
      healthy: this._connected,
      connected: this._connected,
      details: {
        queue: this.queueName,
        queueSize: this.queue?.size() ?? 0,
        processing: this.queue?.processingCount() ?? 0,
        paused: this._paused,
      },
    };
  }

  /**
   * Get the underlying queue (for testing)
   */
  getQueue(): MemoryQueue<T> | null {
    return this.queue;
  }

  /**
   * Get the DLQ (for testing)
   */
  getDLQ(): MemoryQueue<unknown> | null {
    return this.dlq;
  }

  /**
   * Create a wrapped message from stored message
   */
  private createWrappedMessage(stored: StoredMessage<T>): IMessage<T> {
    return createMessage({
      id: stored.id,
      body: stored.body,
      key: stored.key,
      headers: stored.headers,
      timestamp: stored.timestamp,
      deliveryAttempt: stored.deliveryAttempt,
      metadata: {
        provider: 'memory',
        memory: { queueName: this.queueName },
      },
      raw: stored,
      onAck: async () => {
        this.queue?.ack(stored.id);
        this.logger.debug('Message acknowledged', { messageId: stored.id });
      },
      onNack: async (requeue = true) => {
        this.queue?.nack(stored.id, requeue);
        this.logger.debug('Message nacked', { messageId: stored.id, requeue });
      },
    });
  }
}

File: go/memory/consumer.go
package memory

import (
	"context"
	"sync"
	"time"

	"github.com/sns45/anyq/go/core"
)

// Consumer consumes messages from an in-memory queue. It is the reference
// implementation for the retry-strategy hooks: it supports native delayed
// redelivery (setTimeout-style re-enqueue) and native dead-lettering.
type Consumer struct {
	*core.BaseConsumer
	cfg                 Config
	queueName           string
	maxDeliveryAttempts int

	queue *Queue
	dlq   *Queue

	parkMu     sync.Mutex
	parkTimers map[*time.Timer]struct{}
}

// NewConsumer builds a memory consumer.
func NewConsumer(cfg Config) *Consumer {
	if cfg.Driver == "" {
		cfg.Driver = core.DriverMemory
	}
	bc := &core.BaseConsumer{}
	bc.InitBase(cfg.BaseQueueConfig)

	maxDel := 3
	if cfg.DeadLetterQueue != nil && cfg.DeadLetterQueue.MaxDeliveryAttempts > 0 {
		maxDel = cfg.DeadLetterQueue.MaxDeliveryAttempts
	}

	c := &Consumer{
		BaseConsumer:        bc,
		cfg:                 cfg,
		queueName:           cfg.queueName(),
		maxDeliveryAttempts: maxDel,
		parkTimers:          map[*time.Timer]struct{}{},
	}
	bc.Bind(c)
	return c
}

// SupportsNativeDelay: memory schedules delayed re-enqueue via timers.
func (c *Consumer) SupportsNativeDelay() bool { return true }

// Connect attaches to the queue and sets up the DLQ if configured.
func (c *Consumer) Connect(ctx context.Context) error {
	if c.IsConnected() {
		return nil
	}
	if err := c.VerifyParkPolicy(); err != nil {
		return err
	}
	c.queue = GetOrCreateQueue(c.queueName, c.cfg.MaxMessages, c.cfg.MaxAgeMs)

	if dlqCfg := c.cfg.DeadLetterQueue; dlqCfg != nil && dlqCfg.Enabled {
		dlqName := dlqCfg.Destination
		if dlqName == "" {
			dlqName = c.queueName + "-dlq"
		}
		c.dlq = GetOrCreateQueue(dlqName, 0, 0)
	}

	c.SetConnected(true)
	c.Logger.Info("memory consumer connected", map[string]any{"queue": c.queueName})
	return nil
}

// Disconnect stops park timers and marks disconnected.
func (c *Consumer) Disconnect(ctx context.Context) error {
	if !c.IsConnected() {
		return nil
	}
	c.parkMu.Lock()
	for t := range c.parkTimers {
		t.Stop()
	}
	c.parkTimers = map[*time.Timer]struct{}{}
	c.parkMu.Unlock()

	c.SetConnected(false)
	c.Logger.Info("memory consumer disconnected", nil)
	return nil
}

// DeadLetterMessage routes via Queue.DeadLetter so DLQ death headers are set.
func (c *Consumer) DeadLetterMessage(ctx context.Context, msg core.Message, reason string) error {
	stored, _ := msg.Raw().(*StoredMessage)
	if c.queue != nil && c.dlq != nil && stored != nil {
		c.queue.DeadLetter(stored.ID, c.dlq, reason)
		c.Logger.Warn("message dead-lettered", map[string]any{"messageId": msg.ID(), "reason": reason})
		return nil
	}
	c.Logger.Warn("no DLQ configured; dropping message", map[string]any{"messageId": msg.ID(), "reason": reason})
	return msg.Nack(ctx, false)
}

// ParkMessage re-enqueues the message body after delayMs (native delay).
func (c *Consumer) ParkMessage(ctx context.Context, msg core.Message, delayMs int) error {
	if c.queue == nil {
		c.Logger.Warn("parkMessage called without a connected queue", map[string]any{"messageId": msg.ID()})
		return msg.Nack(ctx, false)
	}
	// Remove the in-flight copy so it doesn't linger in processing.
	if stored, ok := msg.Raw().(*StoredMessage); ok {
		c.queue.Ack(stored.ID)
	}

	body := msg.Body()
	key := msg.Key()
	headers := core.MessageHeaders{}
	for k, v := range msg.Headers() {
		headers[k] = v
	}

	var timer *time.Timer
	timer = time.AfterFunc(time.Duration(delayMs)*time.Millisecond, func() {
		c.parkMu.Lock()
		delete(c.parkTimers, timer)
		c.parkMu.Unlock()
		c.queue.Enqueue(body, key, headers)
	})
	c.parkMu.Lock()
	c.parkTimers[timer] = struct{}{}
	c.parkMu.Unlock()

	c.Logger.Debug("message parked for redelivery", map[string]any{"messageId": msg.ID(), "delayMs": delayMs})
	return nil
}

// Subscribe consumes messages until ctx is cancelled or a Fail decision occurs.
func (c *Consumer) Subscribe(ctx context.Context, handler core.Handler, opts *core.SubscribeOptions) error {
	if c.queue == nil {
		return core.NewConnectionError("consumer not connected", nil)
	}
	autoAck := opts.AutoAckEnabled()
	c.Logger.Info("subscribed to queue", map[string]any{"queue": c.queueName})

	for {
		if err := ctx.Err(); err != nil {
			return err
		}
		if c.IsPaused() {
			if !sleepOrDone(ctx, 5*time.Millisecond) {
				return ctx.Err()
			}
			continue
		}
		stored := c.queue.Dequeue()
		if stored == nil {
			if !sleepOrDone(ctx, 5*time.Millisecond) {
				return ctx.Err()
			}
			continue
		}
		msg := c.wrap(stored)
		if err := c.handleOne(ctx, handler, msg, stored, autoAck); err != nil {
			return err
		}
	}
}

func (c *Consumer) handleOne(ctx context.Context, handler core.Handler, msg core.Message, stored *StoredMessage, autoAck bool) error {
	herr := handler(ctx, msg)
	if herr == nil {
		if autoAck {
			return msg.Ack(ctx)
		}
		return nil
	}

	reinvoke := func() error { return handler(ctx, msg) }
	handled, applyErr := c.ApplyStrategy(ctx, msg, herr, reinvoke)
	if applyErr != nil {
		return applyErr // fatal: Fail decision or context cancellation
	}
	if handled {
		return nil
	}

	// Legacy behaviour: DLQ after maxDeliveryAttempts, otherwise requeue.
	c.Logger.Error("error processing message", map[string]any{"messageId": stored.ID, "error": herr.Error()})
	if c.Config.OnError != nil {
		c.Config.OnError(herr)
	}
	if stored.DeliveryAttempt >= c.maxDeliveryAttempts && c.dlq != nil {
		c.queue.DeadLetter(stored.ID, c.dlq, herr.Error())
		c.Logger.Warn("message sent to DLQ", map[string]any{"messageId": stored.ID, "attempts": stored.DeliveryAttempt})
		return nil
	}
	return msg.Nack(ctx, true)
}

// SubscribeBatch consumes message batches until ctx is cancelled.
func (c *Consumer) SubscribeBatch(ctx context.Context, handler core.BatchHandler, opts *core.SubscribeOptions) error {
	if c.queue == nil {
		return core.NewConnectionError("consumer not connected", nil)
	}
	batchSize := 10
	if opts != nil && opts.BatchSize > 0 {
		batchSize = opts.BatchSize
	}
	autoAck := opts.AutoAckEnabled()

	for {
		if err := ctx.Err(); err != nil {
			return err
		}
		if c.IsPaused() {
			if !sleepOrDone(ctx, 5*time.Millisecond) {
				return ctx.Err()
			}
			continue
		}
		storedBatch := c.queue.DequeueBatch(batchSize)
		if len(storedBatch) == 0 {
			if !sleepOrDone(ctx, 5*time.Millisecond) {
				return ctx.Err()
			}
			continue
		}
		msgs := make([]core.Message, len(storedBatch))
		for i, s := range storedBatch {
			msgs[i] = c.wrap(s)
		}

		if herr := handler(ctx, msgs); herr != nil {
			anyUnhandled := false
			for _, m := range msgs {
				mm := m
				reinvoke := func() error { return handler(ctx, []core.Message{mm}) }
				handled, applyErr := c.ApplyStrategy(ctx, mm, herr, reinvoke)
				if applyErr != nil {
					return applyErr
				}
				if !handled {
					anyUnhandled = true
				}
			}
			if anyUnhandled {
				c.Logger.Error("error processing batch", map[string]any{"count": len(msgs), "error": herr.Error()})
				if c.Config.OnError != nil {
					c.Config.OnError(herr)
				}
				for _, m := range msgs {
					_ = m.Nack(ctx, true)
				}
			}
			continue
		}
		if autoAck {
			for _, m := range msgs {
				_ = m.Ack(ctx)
			}
		}
	}
}

// HealthCheck reports consumer health.
func (c *Consumer) HealthCheck(ctx context.Context) (core.HealthStatus, error) {
	connected := c.IsConnected()
	details := map[string]any{"queue": c.queueName, "paused": c.IsPaused()}
	if c.queue != nil {
		details["queueSize"] = c.queue.Size()
		details["processing"] = c.queue.ProcessingCount()
	}
	return core.HealthStatus{Healthy: connected, Connected: connected, Details: details}, nil
}

// Queue returns the underlying queue (for testing).
func (c *Consumer) Queue() *Queue { return c.queue }

// DLQ returns the dead-letter queue, if configured (for testing).
func (c *Consumer) DLQ() *Queue { return c.dlq }

func (c *Consumer) wrap(stored *StoredMessage) core.Message {
	return core.NewMessage(core.MessageParams{
		ID:              stored.ID,
		Body:            stored.Body,
		Key:             stored.Key,
		Headers:         stored.Headers,
		Timestamp:       stored.Timestamp,
		DeliveryAttempt: stored.DeliveryAttempt,
		Metadata: core.ProviderMetadata{
			Provider: core.DriverMemory,
			Memory:   &core.MemoryMetadata{QueueName: c.queueName},
		},
		Raw: stored,
		OnAck: func(context.Context) error {
			c.queue.Ack(stored.ID)
			return nil
		},
		OnNack: func(_ context.Context, requeue bool) error {
			c.queue.Nack(stored.ID, requeue)
			return nil
		},
	})
}

// sleepOrDone sleeps for d, returning false if ctx is cancelled first.
func sleepOrDone(ctx context.Context, d time.Duration) bool {
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-ctx.Done():
		return false
	case <-t.C:
		return true
	}
}

var _ core.Consumer = (*Consumer)(nil)
var _ core.ConsumerHooks = (*Consumer)(nil)
